### ISSUE TITLE
orderedImports should take into account tsconfig.paths

### ISSUE THREAD

**Issue by gligoran** [palantir/tslint#4394]
### Bug Report

- __TSLint version__: 5.12.0
- __TypeScript version__: 3.1.6
- __Running TSLint via__: vscode, CLI, CI, angular CLI

#### TypeScript code being linted

```ts
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { AppRoutingModule } from '@app/app-routing.module';
import { CoreModule } from '@app/core/core.module';
import { SharedModule } from '@app/shared/shared.module';
```

with `tslint.json` configuration:

```json
"ordered-imports": [
    true,
    {
        "grouped-imports": true,
        "import-sources-order": "lowercase-last",
        "module-source-path": "full",
        "named-imports-order": "lowercase-first"
    }
],
```

with `tsconfig.json` configuration:

```json
{
    "compileOnSave": false,
    "compilerOptions": {
        "baseUrl": "./src",
        "declaration": false,
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "importHelpers": true,
        "lib": [
            "es2018",
            "dom",
            "esnext.asynciterable"
        ],
        "module": "commonjs",
        "moduleResolution": "node",
        "noFallthroughCasesInSwitch": true,
        "noImplicitReturns": true,
        "noUnusedLocals": true,
        "outDir": "./dist/out-tsc",
        "paths": {
            "@app/*": [
                "app/*"
            ],
        },
        "removeComments": false,
        "resolveJsonModule": true,
        "skipLibCheck": true,
        "sourceMap": true,
        "strict": true,
        "strictPropertyInitialization": false,
        "target": "es5",
        "typeRoots": [
            "node_modules/@types"
        ]
    }
}
```

#### Actual behavior

Because of the grouped imports, the linter wants to group my named internal imports and my package imports:

```ts
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AppRoutingModule } from '@app/app-routing.module';
import { CoreModule } from '@app/core/core.module';
import { SharedModule } from '@app/shared/shared.module';
```

#### Expected behavior

It would be nice to have the linter leave these along, or even better enforce the split of these so things from node_modules wouldn't mix with things in current codebase.

```ts
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';

import { AppRoutingModule } from '@app/app-routing.module';
import { CoreModule } from '@app/core/core.module';
import { SharedModule } from '@app/shared/shared.module';
```

**Comment by timc13** [comment]
i believe this is a regression bug, NOT a feature request.

The docs state:

> Groups of imports are delineated by blank lines. You can use these to group imports however you like, e.g. by first- vs. third-party or thematically or can you can enforce a grouping of third-party, parent directories and the current directory.

 Indeed this is working in 5.11.0

**Comment by Kerumen** [comment]
The rule `ordered-imports` was bugged in `5.11` (https://github.com/palantir/tslint/issues/4338).

Thanks to https://github.com/palantir/tslint/pull/4374, the rule was fixed, hence "re-enabled". That's why it was "working".

However I agree with the issue, this rule should take into account the `paths` because we don't want these imports to be grouped with the dependencies.

**Comment by johnwiseheart** [comment]
To get to the bottom of this - it seems like there's 2 issues:

The first one is clear to me, which is that if your path starts with an @, TSLint assumes it's a library. In NPM world, starting with an @ implies scope of the package - for instance @blueprint/core and @blueprint/icons, both of which are separate packages under the blueprint banner. I'm not sure where this paradigm of having absolute paths starting with @ came from, but since it seems fairy popular maybe we should look into another way to determine whether an import is a path or a library.

The second issue is less clear to me: @timc what do you mean arbitrary grouping here? Does separating groups by a newline no longer work (this is what the docs describe, if I'm reading them correctly)?

**Comment by timc13** [comment]
According to the docs/spec, the code that @gligoran wrote in the ticket should lint correctly because they are 2 separate groups of imports. I assumed it's breaking currently because the linter is not respecting the group he delineated with the newline.

**Comment by Knagis** [comment]
I have the same issue when enabling this rule - we need to group certain library imports separately because those are our own, even if they are packages, not local imports.

My intention is to add a config option like `localLibraryRegex` that I could specify to match the naming of our packages and thus group them separately between library imports and parent folder imports. (rather for our own purposes I would hardcode the regex).

If this approach seems acceptable I can try to make ir PR-worthy.

**Comment by Knagis** [comment]
https://github.com/palantir/tslint/pull/4134 this seems to be a possible solution for this issue as well

**Comment by adidahiya** [comment]
agreed, I think #4134 is a good solution for this and I'd like to close this issue now that it's merged. open to other feedback (and possibly reopening this issue) after the next TSLint release which includes #4134.
